docs/adrs/00009-re-process-documents.md

```diff
@@ -0,0 +1,133 @@
+allow for easy A/B (green/blue) database setups. Switching the main one to read-only, having the other one run the
+migration.
+
+We could provide an endpoint to the UI, reporting the fact that the system is in read-only mode during a migration.
+
+* 👍 Can fully migrate database (create mandatory field as optional -> re-process -> make mandatory)
+* 👍 Might allow for an out-of-band migration of data, before running the upgrade (even on a staging env)
+* 👍 Would allow to continue serving data while the process is running
+* 👎 Might be tricky to create a combined re-processing of multiple ones
+* 👎 Might block an upgrade if re-processing fails
+
+We do want to support different approaches of this migration. Depending on the needs of the user, the size of the
+data store and the infrastructure used.
+
+### Approach 1
+
+The "lazy" approach, where the user just runs the migration (or the new version of the application with migrations
+enabled). The process will migrate schema and data. This might block the startup for a bit. But would be fast and
+simple for small systems.
+
+### Approach 2
+
+The user uses a green/blue deployment. Switching the application to use green and run migrations against blue. Once
+the migrations are complete, switching back to blue. Green will be read-only and mutable API calls will fail with a 503
+error.
+
+An alternative to this could also be to configure the system first to go into "read-only mode", by using a default
+transaction mode of read-only.
+
+## Consequences
+
+Migrations which do re-process data have to be written in a way, that they can be run and re-run without failing
+during the migration of the schema (e.g. add "if not exist"). In this case, the data migration job can be run
+"out of band" (beforehand) and the data be processed. Then, the actual upgrade and schema migration can run.
+
+
+## Open items
+
+* [ ] Do we want to support downgrades?
+  * I'd say no. Downgrades could also be handled by keeping a snapshot of the original database.
+* [ ] How to handle unparsable or failing documents during migration?
+  * Pass them in as "unsupported"
+* [ ] Add a version number to the document, tracking upgrades
+  * This adds some complexity, but might allow to track the progress and identify upgraded documents. This could also
+    ensure the correct order of applying data migrations out of band.
+
+## Alternative approaches
+
+### Option 2
+
+We create a similar module as for the importer. Running migrations after an upgrade. Accepting that in the meantime,
+we might service inaccurate data.
+
+* 👎 Might serve inaccurate data for a longer time
+* 👎 Can't fully migrate database (new mandatory field won't work)
+* 👍 Upgrade process itself is faster and less complex
+* 👎 Requires some coordination between instances (only one processor at a time, maybe one after the other)
+
+### Option 3
+
+We change ingestion in a way to it is possible to just re-ingest every document. Meaning, we re-ingest from the
```

Comment:
This is something that we do with our internal SCILo system (mentioned in our [kubecon talk](https://docs.google.com/presentation/d/1NU7mAtZqz5UI6hfoi2CKT84oDqF6NbtWN04fpdyeloM/edit)), we run a backfill job that will run based on our source of truth. Keeping the source of truth to one table and a blob store. For our internal infrastructure its not too bad to run it over the entire dataset (in TBs). 

We then read in all the inputs and do a replacement of all the entries in question. We discussed having a table of SoT -> lastProcessedVersion, as a way to keep a list of things that needed to be processed. This would prevent long lasting periods of unavailability as the dataset grows. 

In that way we can do this in a "live" state.